[FLINK-34504][autoscaler] Avoid the parallelism adjustment when the upstream shuffle type doesn't have keyBy

1996fanrui opened the pull request:
## What is the purpose of the change

JobVertexScaler#scale has a optimization: Try to adjust the parallelism such that it divides the number of key groups without a remainder =>  data is evenly spread across subtasks.

It's only useful when the upstream shuffle type has keyBy. We should avoid this optimization when the upstream shuffle type doesn't have keyBy.

## Brief change log

[FLINK-34504][autoscaler] Avoid the parallelism adjustment when the upstream shuffle type doesn't have keyBy

## Verifying this change

- Improve `JobVertexScalerTest#testParallelismComputation`
- Add `JobVertexScalerTest#testParallelismComputationWithKeyedOperator`
- Add `ScalingExecutorTest#testAdjustByMaxParallelism`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no

1996fanrui commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:
```diff
@@ -280,11 +283,14 @@ protected static int scale(
+        // When the shuffle type of vertex data source contains keyBy, we try to adjust the
+        // parallelism such that it divides the number of key groups without a remainder =>
+        // data is evenly spread across subtasks
+        if (hasKeyBy) {
```
Hi, @gyfora @mxm . During I test autoscaler, I found the parallelissm adjustment should not take effect for non-keyed operator.

It's better to include this optimization in 1.8.0. Would you mind helping take a look in your free time? thanks in advanced~

BTW, how about rename  `hasKeyBy` to `isKeyedVertex`?

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java:
```diff
@@ -225,10 +227,14 @@ Map<JobVertexID, ScalingSummary> computeScalingSummary(
                                         (int) metrics.get(ScalingMetric.PARALLELISM).getCurrent();
+
+                                final boolean hasKeyBy =
+                                        jobTopology.get(v).getInputs().containsValue("HASH");
```
It's a bit strange to check string values like `HASH` can we make this an enum so that it's more explicit in the autoscaler code and easier to handle?

1996fanrui commented:
Hi @gyfora , thank you for the quick review!

I have addressed all you comments. For the string shipStrategy, I extracted a `ShipStrategy` enum to refactor it.

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java:
```diff
@@ -225,10 +228,14 @@ Map<JobVertexID, ScalingSummary> computeScalingSummary(
                                         (int) metrics.get(ScalingMetric.PARALLELISM).getCurrent();
+
+                                final boolean hasKeyBy =
+                                        jobTopology.get(v).getInputs().containsValue(HASH);
```
I assume this also means that sources are also not adjusted which can lead to partition imbalance for Kafka topics. We should either keep the adjustments for sources or remove the whole partition number querying/max parallelism logic that we have during collection. 

Overall I think for sources this makes sense and if we want to remove it that should be a separate ticket.

gyfora replied:
This is related to https://issues.apache.org/jira/browse/FLINK-32119

1996fanrui replied:
Thanks a lot for pointing it out. I forgot it before.

I have updated. The source can adjust parallelism for now. Also, I updated parameter names and the comment for `JobVertexScaler#scale` method, for exampe: changing the numKeyGroups to maxParallelism, and changing hasKeyBy to adjustByMaxParallelism.

Because the numKeyGroups and hasKeyBy aren't exactly for source vertex.

gyfora commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java:
```diff
@@ -229,13 +229,14 @@ Map<JobVertexID, ScalingSummary> computeScalingSummary(
                                                 v,
-                                                hasKeyBy,
+                                                adjustByMaxParallelism,
                                                 metrics,
```
Can we please add a single unit test case to cover this core piece of the logic? Just to make sure we don't mess up sources in the future unintentionally

1996fanrui replied:
Good suggestion, added~

1996fanrui replied:
Thank you very much for your review, especially on the weekend. ❤️

gyfora reviewed (approved):
Looks good! Nice improvement :)

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:
```diff
@@ -245,50 +247,69 @@ private boolean detectIneffectiveScaleUp(
         }
     }
 
+    /**
+     * Compute newParallelism according to currentParallelism.
+     *
+     * @param currentParallelism The current parallelism.
+     * @param maxParallelism The max parallelism for job vertices. It's numKeyGroups by default, and
+     *     it's partition number for kafka source vertex.
+     * @param scaleFactor The scale factor.
+     * @param parallelismLowerLimit The parallelism lower limitation in autoscaler option.
+     * @param parallelismUpperLimit The parallelism upper limitation in autoscaler option.
+     * @param adjustByMaxParallelism True means we need to adjust parallelism according to the
+     *     maxParallelism to ensure the keyGroup or partition evenly.
+     */
```
I prefer adding just an explanation of what this method does. There is no additional information which is not already in the method signature. If we want to clarify, e.g. maxParallelism, we can add a dedicated comment for that.

mxm replied:
(I'm suggesting to remove the `@param` lines)

mxm replied:
Awesome, thanks!

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:
```diff
@@ -71,6 +71,7 @@ public JobVertexScaler(AutoScalerEventHandler<KEY, Context> autoScalerEventHandl
     public int computeScaleTargetParallelism(
             Context context,
             JobVertexID vertex,
+            boolean adjustByMaxParallelism,
```
Can we pass-through the ShipStrategy here? That way we decide not based on some boolean but on the relevant data for making the key group adjustment decision. Also makes unit testing more solid.

mxm replied:
Thanks!

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/topology/ShipStrategy.java:
```diff
@@ -0,0 +1,55 @@
+import javax.annotation.Nonnull;
+
+/** The ship strategy between 2 JobVertices. */
+public enum ShipStrategy {
```
Would be nice to reuse Flink's ShipStrategyType if possible.

mxm replied:
Sounds good! Thanks for looking into it. Let's keep it as-is for now.

mxm reviewed (approved):
Thanks Rui. The changes look good! We should keep an eye on the autoscaling behavior after this change because it's actually a breaking change. It might affect the stability of scaling decisions, but I think it should be fine, potentially even improving some situations.

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java:
```diff
@@ -248,25 +250,22 @@ private boolean detectIneffectiveScaleUp(
     @VisibleForTesting
     protected static int scale(
             int currentParallelism,
+            Map<JobVertexID, ShipStrategy> inputs,
```
This could be reduced to just `Collection<ShipStrategy>`, we don't need the vertex.

mxm commented on flink-autoscaler/src/main/java/org/apache/flink/autoscaler/ScalingExecutor.java:
```diff
@@ -229,14 +228,11 @@ Map<JobVertexID, ScalingSummary> computeScalingSummary(
                                                 context,
                                                 v,
-                                                adjustByMaxParallelism,
+                                                jobTopology.get(v).getInputs(),
```
getInputs().values() would suffice.

1996fanrui commented:
Thanks a lot @mxm for the quick review again, updated.
